### DISCUSSION TITLE
Detecting deprecated API versions

### DISCUSSION THREAD

**Discussion by djfinnoy** [argoproj/argo-cd#5867] [General]
We would like to be alerted when a deprecated API is detected within our k8s cluster.
Found [this tool](https://github.com/doitintl/kube-no-trouble), but I doubt it will work with resources deployed via ArgoCD (it only detects api versions from resources deployed via Helm or kubectl).

I'm wondering if there are any straight-forward ways to figure out what api versions are in use for resources deployed via ArgoCD?

**Comment by jannfis** [comment]
Since Argo CD under the hood uses same mechanisms as `kubectl`, I think you might have some success with `kubent`. Have you actually tried it?

**Reply by djfinnoy** [reply]
Not yet, will report back after I have

**Comment by der-eismann** [comment]
Were you able to find a solution? I try to keep track, but I'm afraid of the day when we will upgrade Kubernetes and suddenly some resources can't be applied. Would be great to see the deprecations somewhere in ArgoCD.

**Reply by djfinnoy** [reply]
It's possible to detect deprecated API versions with `kubent`, even for resources deployed via ArgoCD.

However, it was a bit of a pain to create a cronjob that utilizes `kubent`, and monitoring rules to create an alert when something is deprecated. In the end we just decided to use `kubent` manually.

We adopted a green-blue cluster deploy pattern to guard against the day when an update will cause downtime. If there are dangerous changes in some upgrade, we roll these changes out in a new cluster, and scale down the old cluster when confident that everything still works.
